core/kernel/persistence/smoothsql/class.Resource.php

```diff
@@ -102,6 +103,135 @@ public function getTypes(core_kernel_classes_Resource $resource)
         return (array) $returnValue;
     }
 
+    public function getParentClassesIds(string $resourceUri): array
+    {
+        $query = <<<'SQL'
+WITH RECURSIVE statements_tree AS (
+    SELECT
+        r.object
+    FROM statements r
+    WHERE r.subject = ?
+      AND r.predicate IN (?, ?)
+      AND r.object != ?
+    UNION ALL
+    SELECT
+        s.object
+    FROM statements s
+        JOIN statements_tree st
+            ON s.subject = st.object
+                   AND s.predicate IN (?, ?)
+                   AND s.object NOT IN (?, ?, ?, ?)
+)
+SELECT object FROM statements_tree;
+SQL;
+
+        $statement = $this->getPersistence()->query(
+            $query,
+            [
+                $resourceUri,
+                OntologyRdfs::RDFS_SUBCLASSOF,
+                OntologyRdf::RDF_TYPE,
+                'http://www.tao.lu/Ontologies/TAO.rdf#AssessmentContentObject',
+                OntologyRdfs::RDFS_SUBCLASSOF,
+                OntologyRdf::RDF_TYPE,
+                'http://www.tao.lu/Ontologies/TAO.rdf#AssessmentContentObject',
+                'http://www.tao.lu/Ontologies/TAO.rdf#TAOObject',
+                'http://www.tao.lu/Ontologies/generis.rdf#generis_Ressource',
+                'http://www.w3.org/2000/01/rdf-schema#Resource',
+            ]
+        );
+
+        return array_column($statement->fetchAll(), 'object');
+    }
+
+    /**
+     * @param array $classIds
+     * @return array [
+     *     '{classUri}' => [
+     *         '{resourceUri_1}',
+     *         '{resourceUri_N...}',
+     *     ]
+     * ]
+     */
+    public function getClassesResourceIds(array $classIds): array
+    {
+        $inQuery = implode(',', array_fill(0, count($classIds), '?'));
+        $query = "SELECT subject, object FROM statements WHERE predicate IN (?, ?) AND object IN ($inQuery)";
+
+        $statement = $this->getPersistence()->query(
+            $query,
+            [
+                OntologyRdfs::RDFS_SUBCLASSOF,
+                OntologyRdf::RDF_TYPE,
+                ...$classIds,
+            ]
+        );
+
+        $results = $statement->fetchAll();
+        $resourceIds = [];
+
+        foreach ($classIds as $classId) {
+            /**
+             * @TODO FIXME Better filter it instead of go through all results every time
```

Comment:
- [ ] TODO WE probably can avoid loop through all the records every time

cc: @shpran